server/src/main/java/org/elasticsearch/node/Node.java

```diff
@@ -705,6 +705,8 @@ public Node start() throws NodeValidationException {
         assert localNodeFactory.getNode() != null;
         assert transportService.getLocalNode().equals(localNodeFactory.getNode())
             : "transportService has a different local node than the factory provided";
+        onTransportServiceStarted();
```

Comment:
I think you can avoid adding these changes here in Node (and MockNode), but instead after initializing the node (but before starting it), call `node.injector().getInstance(TransportService.class)` to get the `TransportService` and then register a lifecycle listener on that (`addLifecycleListener`) which implements `afterStart`.